# Hyperparameter Tuning using HyperDrive

Import Dependencies. In the cell below, import all the dependencies that you will need to complete the project.

In [1]:
from azureml.core import Workspace, Experiment, Environment, ScriptRunConfig
from azureml.data.dataset_factory import TabularDatasetFactory as tdf
from azureml.core.compute import ComputeTarget, AmlCompute
from azureml.core.compute_target import ComputeTargetException
from azureml.widgets import RunDetails
from azureml.train.sklearn import SKLearn
from azureml.train.hyperdrive.run import PrimaryMetricGoal
from azureml.train.hyperdrive.policy import BanditPolicy, MedianStoppingPolicy
from azureml.train.hyperdrive.sampling import RandomParameterSampling
from azureml.train.hyperdrive.runconfig import HyperDriveConfig
from azureml.train.hyperdrive.parameter_expressions import uniform, choice
import os
import joblib

## Set Up

### Load Workspace Elements, Create an Experiment

In [2]:
ws = Workspace.from_config()

# choose a name for experiment
experiment_name = 'MS-Malware-Hyper'
project_folder = '.'

experiment=Experiment(ws, experiment_name)

print('Workspace name: ' + ws.name, 
      'Azure region: ' + ws.location, 
      'Subscription id: ' + ws.subscription_id, 
      'Resource group: ' + ws.resource_group, sep = '\n')

Workspace name: quick-starts-ws-133879
Azure region: southcentralus
Subscription id: 6971f5ac-8af1-446e-8034-05acea24681f
Resource group: aml-quickstarts-133879


### Create/Get Compute Cluster

In [3]:
cpu_cluster_name = "malware-compute"

# if cluster already exists, use it
try:
    cpu_cluster = ComputeTarget(workspace=ws, name=cpu_cluster_name)
    print('Cluster {} exists. Will use this cluster.'.format(cpu_cluster_name))
except ComputeTargetException:
    compute_config = AmlCompute.provisioning_configuration(vm_size='STANDARD_D2_V2', max_nodes=4)
    cpu_cluster = ComputeTarget.create(ws, cpu_cluster_name, compute_config)
    
    cpu_cluster.wait_for_completion(show_output=True)

Cluster malware-compute exists. Will use this cluster.


### Create an Environment

In [4]:
%%writefile hyperdrive_dependencies.yml

dependencies:
- python=3.6.2
- scikit-learn
- pandas 
- numpy
- pip:
    - azureml-defaults
    - xgboost

Writing hyperdrive_dependencies.yml


In [5]:
hyperdrive_env = Environment.from_conda_specification(name = 'hyperdrive-env', file_path = './hyperdrive_dependencies.yml')

## Data Set

See the "automl.ipynb" Notebook for Data Set explanation

In [6]:
# Load Data set
data_path = 'https://raw.githubusercontent.com/tybyers/AZMLND_projects/capstone/capstone/data/train_1_10k.csv'
dataset = tdf.from_delimited_files(path=data_path)
dataset.to_pandas_dataframe().head()

ProductName EngineVersion       AppVersion  AvSigVersion  IsBeta  \
0  win8defender   1.1.15100.1  4.18.1807.18075  1.273.1735.0       0   
1  win8defender   1.1.14600.4     4.13.17134.1    1.263.48.0       0   
2  win8defender   1.1.15100.1  4.18.1807.18075  1.273.1341.0       0   
3  win8defender   1.1.15100.1  4.18.1807.18075  1.273.1527.0       0   
4  win8defender   1.1.15100.1  4.18.1807.18075  1.273.1379.0       0   

  RtpStateBitfield  IsSxsPassiveMode DefaultBrowsersIdentifier  \
0                7                 0                        NA   
1                7                 0                        NA   
2                7                 0                        NA   
3                7                 0                        NA   
4                7                 0                        NA   

   AVProductStatesIdentifier  AVProductsInstalled  ...  \
0                    53447.0                  1.0  ...   
1                    53447.0                  1.0  ...   
2                    53447.0                  1.0  ...   
3                    53447.0                  1.0  ...   
4                    53447.0                  1.0  ...   

   Census_FirmwareVersionIdentifier  Census_IsSecureBootEnabled  \
0                             36144                           0   
1                             57858                           0   
2                             52682                           0   
3                             20050                           0   
4                             19844                           0   

   Census_IsWIMBootEnabled Census_IsVirtualDevice Census_IsTouchEnabled  \
0                       NA                      0                     0   
1                       NA                      0                     0   
2                       NA                      0                     0   
3                       NA                      0                     0   
4                        0                      0                     0   

   Census_IsPenCapable  Census_IsAlwaysOnAlwaysConnectedCapable Wdft_IsGamer  \
0                    0                                      0.0            0   
1                    0                                      0.0            0   
2                    0                                      0.0            0   
3                    0                                      0.0            0   
4                    0                                      0.0            0   

  Wdft_RegionIdentifier HasDetections  
0                    10             0  
1                     8             0  
2                     3             0  
3                     3             1  
4                     1             1  

[5 rows x 82 columns]

## Hyperdrive Configuration

For the Hyperparameter tuning experiment, done via HyperDrive, we opted for tuning an XGBoost model. We did this because, of the top 10 performing AutoML models, 4 of them were XGBoost Classifiers. We are unsure how much hyperparameter tuning, if any, happens with AutoML, so we wanted to see if we could do any better than the AutoML XGBoost.

### Param Sampling

We opted to tune the following hyperparameters because, based on our experience using XGBoost in the past, these are some of the highest-leverage hyperparameters to tune:

```python
param_sampling = RandomParameterSampling({'--max_depth': choice(range(2,11)),
                                         '--n_estimators': choice(25, 50, 100, 250, 500, 750, 1000),
                                         '--learning_rate': uniform(0, 1.0)})`
```
* `max_depth` - Maximum depth of a tree. Increasing this value will make the model more complex and more likely to overfit.
* `n_estimators` - Number of boosting rounds. Trains faster when smaller; too large can overfit.
* `learning_rate` - Step size shrinkage used in update to prevents overfitting. Range: [0, 1]

### Other Configuration Settings 

* `early_termination_policy = BanditPolicy` - This is an early termination policy that terminates any runs where the metric is not performing well compared to the best-performing run. The benefit of this policy is that especially for long-running trials it will terminate early if the accuracy metric is too far off the best run so far, and in that way it will speed up the overall compute.  
* `environment=hyperdrive_env` -- Explained below  
* `primary_metric_name='Accuracy'` -- Opted for Accuracy here to be consistent with the AutoML experiment.  
* `primary_metric_goal=PrimaryMetricGoal.MAXIMIZE` -- Obviously we want the highest accuracy and not the lowest :). 
* `max_total_runs=100` -- XGBoost can train fairly fast, so let's see if we get any real outstanding hits. Even so, limit to only 100 runs.  
* `max_concurrent_runs=4` -- Run up to 4 models in parallel.  

In [7]:
# TODO: Create an early termination policy. This is not required if you are using Bayesian sampling.
early_termination_policy = BanditPolicy(evaluation_interval=2, slack_factor=0.1)

#TODO: Create the different params that you will be using during training
param_sampling = RandomParameterSampling({'--max_depth': choice(range(2,11)),
                                         '--n_estimators': choice(25, 50, 100, 250, 500, 750, 1000),
                                         '--learning_rate': uniform(0, 1.0)})

src = ScriptRunConfig(source_directory=project_folder,
                      script='xgbtrain.py',
#                      arguments=['--kernel', 'linear', '--penalty', 1.0],
                      compute_target=cpu_cluster,
                      environment=hyperdrive_env)

hyperdrive_config = HyperDriveConfig(run_config=src,
                                    hyperparameter_sampling=param_sampling,
                                    policy=early_termination_policy,
                                    primary_metric_name='Accuracy',
                                    primary_metric_goal=PrimaryMetricGoal.MAXIMIZE,
                                    max_total_runs=100,
                                    max_concurrent_runs=4)

In [8]:
#TODO: Submit your experiment
hyperdrive_run = experiment.submit(hyperdrive_config)

## Run Details

OPTIONAL: Write about the different models trained and their performance. Why do you think some models did better than others?

TODO: In the cell below, use the `RunDetails` widget to show the different experiments.

In [9]:
RunDetails(hyperdrive_run).show()

_HyperDriveWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO'…

In [10]:
hyperdrive_run.wait_for_completion(show_output=True)

RunId: HD_3168198a-ecc3-4d37-918b-7964abf6ec22
Web View: https://ml.azure.com/experiments/MS-Malware-Hyper/runs/HD_3168198a-ecc3-4d37-918b-7964abf6ec22?wsid=/subscriptions/6971f5ac-8af1-446e-8034-05acea24681f/resourcegroups/aml-quickstarts-133879/workspaces/quick-starts-ws-133879

Streaming azureml-logs/hyperdrive.txt

"<START>[2021-01-07T19:12:49.592047][GENERATOR][INFO]Trying to sample '4' jobs from the hyperparameter space<END>\n""<START>[2021-01-07T19:12:48.739627][API][INFO]Experiment created<END>\n"<START>[2021-01-07T19:12:50.1242844Z][SCHEDULER][INFO]The execution environment is being prepared. Please be patient as it can take a few minutes.<END>"<START>[2021-01-07T19:12:49.904766][GENERATOR][INFO]Successfully sampled '4' jobs, they will soon be submitted to the execution target.<END>\n"

Execution Summary
RunId: HD_3168198a-ecc3-4d37-918b-7964abf6ec22
Web View: https://ml.azure.com/experiments/MS-Malware-Hyper/runs/HD_3168198a-ecc3-4d37-918b-7964abf6ec22?wsid=/subscriptions/697

{'runId': 'HD_3168198a-ecc3-4d37-918b-7964abf6ec22',
 'target': 'malware-compute',
 'status': 'Completed',
 'startTimeUtc': '2021-01-07T19:12:48.488583Z',
 'endTimeUtc': '2021-01-07T20:36:15.897339Z',
 'properties': {'primary_metric_config': '{"name": "Accuracy", "goal": "maximize"}',
  'resume_from': 'null',
  'runTemplate': 'HyperDrive',
  'azureml.runsource': 'hyperdrive',
  'platform': 'AML',
  'ContentSnapshotId': '99c3333a-a0b2-4380-aaf4-69b97d73abc4',
  'score': '0.626',
  'best_child_run_id': 'HD_3168198a-ecc3-4d37-918b-7964abf6ec22_55',
  'best_metric_status': 'Succeeded'},
 'inputDatasets': [],
 'outputDatasets': [],
 'logFiles': {'azureml-logs/hyperdrive.txt': 'https://mlstrg133879.blob.core.windows.net/azureml/ExperimentRun/dcid.HD_3168198a-ecc3-4d37-918b-7964abf6ec22/azureml-logs/hyperdrive.txt?sv=2019-02-02&sr=b&sig=9e3MZEQHWG3USzhCMisgYX9C%2BzQ%2Bv2z15sY7u81zSfQ%3D&st=2021-01-07T20%3A27%3A08Z&se=2021-01-08T04%3A37%3A08Z&sp=r'}}

In [11]:
assert(hyperdrive_run.get_status() == "Completed")

## Best Model

TODO: In the cell below, get the best model from the hyperdrive experiments and display all the properties of the model.

In [12]:
best_run = hyperdrive_run.get_best_run_by_primary_metric()
best_run.get_metrics()

{'max_depth:': 2,
 'n_estimators:': 100,
 'learning_rate:': 0.1897631244476532,
 'Accuracy': 0.626}

In [13]:
best_run.properties

{'_azureml.ComputeTargetType': 'amlcompute',
 'ContentSnapshotId': '99c3333a-a0b2-4380-aaf4-69b97d73abc4',
 'ProcessInfoFile': 'azureml-logs/process_info.json',
 'ProcessStatusFile': 'azureml-logs/process_status.json'}

Save the best model

In [14]:
print(best_run.get_details()['runDefinition'])

{'script': 'xgbtrain.py', 'command': '', 'useAbsolutePath': False, 'arguments': ['--learning_rate', '0.1897631244476532', '--max_depth', '2', '--n_estimators', '100'], 'sourceDirectoryDataStore': None, 'framework': 'Python', 'communicator': 'None', 'target': 'malware-compute', 'dataReferences': {}, 'data': {}, 'outputData': {}, 'jobName': None, 'maxRunDurationSeconds': 2592000, 'nodeCount': 1, 'priority': None, 'credentialPassthrough': False, 'environment': {'name': 'hyperdrive-env', 'version': 'Autosave_2021-01-07T19:12:48Z_6d978402', 'python': {'interpreterPath': 'python', 'userManagedDependencies': False, 'condaDependencies': {'dependencies': ['python=3.6.2', 'scikit-learn', 'pandas', 'numpy', {'pip': ['azureml-defaults', 'xgboost']}], 'name': 'azureml_523fb8d0d2520274d0060eaa29065899'}, 'baseCondaEnvironment': None}, 'environmentVariables': {'EXAMPLE_ENV_VAR': 'EXAMPLE_VALUE'}, 'docker': {'baseImage': 'mcr.microsoft.com/azureml/intelmpi2018.3-ubuntu16.04:20200821.v1', 'platform': {

In [15]:
print(best_run.get_file_names())

['azureml-logs/55_azureml-execution-tvmps_acf603eb6bb5ba18b69f349603579855633b37da738e34748ab20f76e4b686b7_d.txt', 'azureml-logs/65_job_prep-tvmps_acf603eb6bb5ba18b69f349603579855633b37da738e34748ab20f76e4b686b7_d.txt', 'azureml-logs/70_driver_log.txt', 'azureml-logs/75_job_post-tvmps_acf603eb6bb5ba18b69f349603579855633b37da738e34748ab20f76e4b686b7_d.txt', 'azureml-logs/process_info.json', 'azureml-logs/process_status.json', 'logs/azureml/101_azureml.log', 'logs/azureml/dataprep/backgroundProcess.log', 'logs/azureml/dataprep/backgroundProcess_Telemetry.log', 'logs/azureml/dataprep/engine_spans_l_176dbe34-b82b-42bc-b8fe-ca55d359239e.jsonl', 'logs/azureml/dataprep/python_span_l_176dbe34-b82b-42bc-b8fe-ca55d359239e.jsonl', 'logs/azureml/job_prep_azureml.log', 'logs/azureml/job_release_azureml.log', 'outputs/model.joblib']


In [16]:
model = best_run.register_model(model_name = 'best_hyperdrive', model_path='outputs/model.joblib')
print(model.name, model.id, model.version, sep='\t')

best_hyperdrive	best_hyperdrive:1	1


## Model Deployment

We opted to deploy the AutoML model because it had a higher accuracy -- 0.638 compared to the best accuracy here of 0.626. Please see the automl.ipynb notebook for details of that deployment.